src/snowflake/snowpark/mock/plan.py

```diff
@@ -982,7 +1072,268 @@ def calculate_expression(
+        elif isinstance(window_function, (Lead, Lag)):
+            offset = window_function.offset * (
+                1 if isinstance(window_function, Lead) else -1
+            )
+            ignore_nulls = window_function.ignore_nulls
+            res_cols = []
+            for current_row, w in zip(res_index, windows):
+                row_idx = list(w.index).index(
+                    current_row
+                )  # the row's 0-base index in the window
+                offset_idx = row_idx + offset
+                if offset_idx < 0 or offset_idx >= len(w):
+                    res_cols.append(
+                        calculate_expression(
+                            window_function.default,
+                            w,
+                            analyzer,
+                            expr_to_alias,
+                            keep_literal=True,
+                        )
+                    )
+                elif not ignore_nulls or offset == 0:
+                    res_cols.append(
+                        calculate_expression(
+                            window_function.expr,
+                            w.iloc[offset_idx],
+                            analyzer,
+                            expr_to_alias,
+                            keep_literal=True,
+                        )
+                    )
+                else:
+                    # skip rows where expr is NULL
+                    delta = 1 if offset > 0 else -1
+                    cur_idx = row_idx + delta
+                    cur_count = 0
+                    while 0 <= cur_idx < len(w):
+                        target_expr = calculate_expression(
+                            window_function.expr,
+                            w.iloc[cur_idx],
+                            analyzer,
+                            expr_to_alias,
+                            keep_literal=True,
+                        )
+                        if target_expr is not None:
+                            cur_count += 1
+                            if cur_count == abs(offset):
+                                break
+                        cur_idx += delta
+                    if cur_idx < 0 or cur_idx >= len(w):
+                        res_cols.append(
+                            calculate_expression(
+                                window_function.default,
+                                w,
+                                analyzer,
+                                expr_to_alias,
+                                keep_literal=True,
+                            )
+                        )
+                    else:
+                        res_cols.append(target_expr)
+
+            res_col = ColumnEmulator(
+                data=res_cols, dtype=object
+            )  # dtype=object prevents implicit converting None to Nan
+            res_col.index = res_index
+            return res_col.sort_index()
+        elif isinstance(window_function, FirstValue):
```

Comment:
nit: it feels that FirstValue and LastValue calculation are sharing lots of logic
except for the `w.iloc[0]` vs `w.iloc[len(w) - 1]` and `for cur_idx in range(len(w)):` vs `for cur_idx in range(len(w) - 1, -1, -1):`.

Reply:
This is very true. I am planning to refactor this part in a separate PR so all window function definitions can be found in `mock/functions.py`.

Reply:
let's create a JIRA or log it in an existing jira so that we don't forget?